concordium-std/src/lib.rs

```diff
@@ -207,9 +225,89 @@
+//! [test_infrastructure] and are not needed in the new testing library.
+//! The primary of these traits are [`HasHost`], [`HasStateApi`],
+//! [`HasInitContext`], and [`HasReceiveContext`].
+//!
+//! ## Migration guide
+//! To migrate your contract and its tests to the new testing library, you need
+//! to do the following two steps:
+//!
+//! 1. Replace the usage of deprecated traits with their concrete alternatives
+//! and remove generics.
+//!
+//!    For init methods:
+//!    ```no_run
+//!    # use concordium_std::*;
+//!    #
+//!    # type State = ();
+//!    #
+//!    /// Before
+//!    #[init(contract = "contract_before")]
+//!    fn init_before<S: HasStateApi>(
+//!        ctx: &impl HasInitContext,
+//!        state_builder: &mut StateBuilder<S>,
+//!    ) -> InitResult<State> { todo!() }
+//!
+//!    /// After
+//!    #[init(contract = "contract_after")]
+//!    fn init_after(                        // `<S: HasStateApi>` removed
+//!        ctx: &InitContext,                // `impl` and `Has` removed
+//!        state_builder: &mut StateBuilder, // `<S>` removed
+//!    ) -> InitResult<State> { todo!() }
+//!    ```
+//!    For receive methods:
+//!    ```no_run
+//!    # use concordium_std::*;
+//!    #
+//!    # type State = ();
+//!    # type MyReturnValue = ();
+//!    #
+//!    # #[init(contract = "my_contract")]
+//!    # fn contract_init(                    // `<S: HasStateApi>` removed
+//!    #    ctx: &InitContext,                // `impl` and `Has` removed
+//!    #    state_builder: &mut StateBuilder, // `<S>` removed
+//!    # ) -> InitResult<State> { todo!() }
+//!    /// Before
+//!    #[receive(contract = "my_contract", name = "my_receive")]
+//!    fn receive_before<S: HasStateApi>(
+//!        ctx: &impl HasReceiveContext,
+//!        host: &impl HasHost<State, StateApiType = S>,
+//!    ) -> ReceiveResult<MyReturnValue> { todo!() }
+//!    
+//!    /// After
+//!    #[receive(contract = "my_contract", name = "my_receive")]
+//!    fn receive_after(           // `<S: HasStateApi>` removed
+//!        ctx: &ReceiveContext,   // `impl` and `Has` removed
+//!        host: &Host<State>,     // `impl Has` and `, StateApiType = S removed
+//!    ) -> ReceiveResult<MyReturnValue> { todo!() }
+//!    ```
+//!
+//!    If you use logging, crypto-primitives, or similar, you must also also
+//! replace those uses of traits with concrete types. E.g. replacing `&mut impl
+//! HasLogger` with `&mut Logger`.
+//!
+//! 2. Migrate your tests to use the new testing library.
+//!
+//!    For an introduction to the library, see our [guide](https://developer.concordium.software/en/mainnet/smart-contracts/guides/integration-test-contract.html).
+//!
+//!    If you follow our [recommended structure](https://developer.concordium.software/en/mainnet/smart-contracts/best-practices/development.html#recommended-structure) in your contract,
+//!    then you can easily keep unit tests that only call methods directly on
```

Comment:
It is not completely clear what you mean here. Do you suggest that people can keep the unit tests when the methods directly call on the state struct but should add the `#[allow(deprecated)]` to the test case for now?

Reply:
Thanks for pointing this out, it now reads:

```
//!    If you follow our [recommended structure](https://developer.concordium.software/en/mainnet/smart-contracts/best-practices/development.html#recommended-structure) in your contract,
//!    then you have a mix of unit and integrations tests:
//!    - Unit tests that call methods directly on your state struct (without any
//!      init/receive calls)
//!    - Integration tests that call the init and receive methods
//!
//! If you do not want to migrate your contract and tests yet, then you can add
//! the `#[allow(deprecated)]` attribute to your test modules to avoid the
//! deprecation warnings.
```